### PR TITLE
Add new section about writing Go policies

### PR DESCRIPTION
**Pull request by flavio**
This is a major rewrite of the old documentation that explained how to write Go policies.

I like how I broke down everything into dedicated sections. I'll make a PR to adopt a similar structure also for the Rust instructions.

### CODE REVIEW

File: src/SUMMARY.md [hunk 1]
```diff
@@ -5,10 +5,16 @@
 - [Testing Policies](./testing-policies.md)
 - [Writing Policies](./writing-policies/index.md)
   - [Rust](./writing-policies/rust.md)
+  - [Go](./writing-policies/go/01-intro.md)
+    - [Create a new policy](./writing-policies/go/02-scaffold.md)
+    - [Define policy settings](./writing-policies/go/03-policy-settings.md)
+    - [Write the validation logic](./writing-policies/go/04-valitation.md)
```
Comment (ereslibre) [inline src/SUMMARY.md:11]:
```suggestion
    - [Write the validation logic](./writing-policies/go/04-validation.md)
```

File: src/writing-policies/go/04-valitation.md [hunk 2]
```diff
@@ -0,0 +1,351 @@
+# Writing the validation logic
```
Comment (ereslibre) [inline src/writing-policies/go/04-valitation.md:1]: A new file will appear when the SUMMARY.md file is fixed. Make sure to do further changes on the file with the right name, and remove the old one.

File: src/writing-policies/go/01-intro.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+TinyGo doesn't yet support all the Go features (see [here](https://tinygo.org/lang-support/)
+to see the current project status). Currently its biggest limitation
+is the lack of a fully supported `reflect` package. That leads to the inability to use
+the `encoding/json` package against structures and user defined types.
+
+Kubewarden policies need to process JSON data like the policy settings and
+the actual request received by Kubernetes.
+
+Depite TinyGo's current limitations, it still easy and doable to write Kubewarden policies
```
Comment (ereslibre) [inline src/writing-policies/go/01-intro.md:34]:
```suggestion
Despite TinyGo's current limitations, it's still easy and doable to write Kubewarden policies
```

File: src/writing-policies/go/01-intro.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+with it.
+
+## Tooling
+
+Writing Kubewarden Kubewarden's policies requires a version of TinyGo major than `0.17.0`.
```
Comment (ereslibre) [inline src/writing-policies/go/01-intro.md:39]:
```suggestion
Writing Kubewarden policies requires a version of TinyGo greater than `0.17.0`.
```

File: src/writing-policies/go/01-intro.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+* [Kubewarden Go SDK](https://github.com/kubewarden/policy-sdk-go): provides a series of
+  structures and functions that reduce the amount of code to write. It also provides test helpers.
+* [gjson](https://github.com/tidwall/gjson): provides a powerful query language that allows
+  quick navigation of JSON documents and data retrieval. This library doesn't use the
+  `encoding/json` package provided by Go's stdlib, hence it's usable with TinyGo.
+* [sjson](https://github.com/tidwall/sjson): can be used to mutate JSON documents.
+  This library doesn't use the `encoding/json` package provided by Go's stdlib, hence it's usable with TinyGo.
+* [mapset](https://github.com/deckarep/golang-set): provides a Go implementation of the
+  [Set](https://en.wikipedia.org/wiki/Set_(abstract_data_type))
+  data types. This library significantly reduces the amount of code to be written,
```
Comment (ereslibre) [inline src/writing-policies/go/01-intro.md:52]:
```suggestion
  data structure. This library significantly reduces the amount of code to be written,
```

?

File: src/writing-policies/go/01-intro.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  of policies.
+
+Last but not least, the Kubewarden project provides a
+[template Go policy project](https://github.com/kubewarden/go-policy-template)
+that can be used to quickly create Kubewarden policies written in Go.
+
+## Getting TinyGo dependencies
+
+The easiest way to get TinyGo is by using the upstream container images.
+Official release can be found [here](https://hub.docker.com/r/tinygo/tinygo), while
```
Comment (ereslibre) [inline src/writing-policies/go/01-intro.md:63]:
```suggestion
Official releases can be found [here](https://hub.docker.com/r/tinygo/tinygo), while
```

File: src/writing-policies/go/01-intro.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+builds from the development branch are automatically pushed
+[here](https://hub.docker.com/r/tinygo/tinygo-dev).
+
+If needed, checkout TinyGo's [getting started](https://tinygo.org/getting-started/) page for
+more information.
+
+> **Note well:** Kubewarden's requires code that is available only on the development branch. This will
+> be solved once TinyGo release `0.17.0`. In the meantime we will use the container image
```
Comment (ereslibre) [inline src/writing-policies/go/01-intro.md:71]:
```suggestion
> be solved once TinyGo `0.17.0` is released. In the meantime we will use the container image
```

File: src/writing-policies/go/02-scaffold.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+To summarize, the policy settings will look like that:
+
+```yaml
+# List of labels that cannot be used
+denied_labels:
+- foo
+- bar
+
+# Labels that are validate with user-defined RegExp
```
Comment (ereslibre) [inline src/writing-policies/go/02-scaffold.md:18]:
```suggestion
# Labels that are validated with user-defined RegExp
```

File: src/writing-policies/go/02-scaffold.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+# Failing to comply with the RegExp resuls in the object
+# being rejected
```
Comment (ereslibre) [inline src/writing-policies/go/02-scaffold.md:20]: This comment does not seem to hold together. Different sentences?

File: src/writing-policies/go/04-valitation.md [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+	"encoding/json"
+	"testing"
+
+	"github.com/deckarep/golang-set"
+	kubewarden_testing "github.com/kubewarden/policy-sdk-go/testing"
+)
+```
+
+The first unit test will ensure that having no user settings leads
+to the request to be approved:
```
Comment (ereslibre) [inline src/writing-policies/go/04-valitation.md:113]: Should we use the same terminology? `Accepted` and in other instances of approve later on.

File: src/writing-policies/go/05-e2e-tests.md [hunk 11]
```diff
@@ -0,0 +1,212 @@
+# End-to-end testing
```
Comment (ereslibre) [inline src/writing-policies/go/05-e2e-tests.md:1]: For later: I have the impression we can reuse this section for all languages. On each language we explain how to obtain the Wasm binary, and from that point on, e2e can be achieved with the same strategy for all of them.

File: src/writing-policies/go/06-automate.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+We can automate the execution of the unit tests and of the end-to-end tests.
+
+This is working out of the box thanks to these files:
+
+  * `.github/workflows/e2e-tests.yml`
+  * `.github/workflows/unit-tests.yml`
+
+## Release
+
+The scaffolded projected created a `.github/workflows/release.yml.template`.
```
Comment (ereslibre) [inline src/writing-policies/go/06-automate.md:22]:
```suggestion
The scaffolded project created a `.github/workflows/release.yml.template`.
```

File: src/writing-policies/go/06-automate.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+The pipeline will act differently based on the commit that triggered its execution.
+
+Regular commits will lead to the creation of an OCI artifact called `<policy-name>:latest`.
+No GitHub Release will be created for this commits.
+
+On the other hand, creating a tag that matches the `v*` pattern, will lead
+to:
+
+1. Creation of an OCI artifact called `<policy-name>:<tag>`.
+1. Creation of an GitHub Release named `Release <full tag name>`. The release
```
Comment (ereslibre) [inline src/writing-policies/go/06-automate.md:43]:
```suggestion
1. Creation of a GitHub Release named `Release <full tag name>`. The release
```

File: src/writing-policies/go/06-automate.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  will include the following assets: the source code of the policy, the WebAssembly
```
Comment (ereslibre) [inline src/writing-policies/go/06-automate.md:44]:
```suggestion
  will include the following assets: the source code of the policy and the WebAssembly
```

File: src/writing-policies/go/07-distribute.md [hunk 15]
```diff
@@ -0,0 +1,18 @@
+# Distribute policy
+
+Congratulations for having made that far 🎉🎉🎉
```
Comment (ereslibre) [inline src/writing-policies/go/07-distribute.md:3]:
```suggestion
Congratulations for having made this far 🎉🎉🎉
```

**Review by ereslibre — CHANGES_REQUESTED** [review]
I did some quick scanning for typos and things that don't hold. I haven't looked deep into the examples, but I think after this round we can move forward and merge; then improve later.

Thank you @flavio!

File: src/SUMMARY.md (hunk 1 above)
Comment (ereslibre) [reply src/SUMMARY.md:11]: Note well: renaming will create a new file and leave the old one around.

Comment (flavio) [reply src/SUMMARY.md:11]: I fixed that manually. Thanks

File: src/writing-policies/go/02-scaffold.md (hunk 9 above)
Comment (flavio) [reply src/writing-policies/go/02-scaffold.md:20]: solved, thanks

File: src/writing-policies/go/04-valitation.md (hunk 10 above)
Comment (flavio) [reply src/writing-policies/go/04-valitation.md:113]: I fixed that across of all the file (also inside of the code)

**Comment by flavio** [comment]
I applied all the suggestions, fixed the other things you reported and then squashed everything together with you as co-author.

I'm going to merge it

**Pull request merged**